src/transport/SecureSessionMgr.cpp

```diff
@@ -78,150 +78,165 @@ CHIP_ERROR SecureSessionMgr::Init(NodeId localNodeId, System::Layer * systemLaye
 
     mTransportMgr->SetSecureSessionMgr(this);
 
-exit:
-    return err;
+    return CHIP_NO_ERROR;
 }
 
 CHIP_ERROR SecureSessionMgr::SendMessage(NodeId peerNodeId, System::PacketBufferHandle msgBuf)
 {
-    PayloadHeader payloadHeader;
-
-    return SendMessage(payloadHeader, peerNodeId, std::move(msgBuf));
+    PayloadHeader unusedPayloadHeader;
+    return SendMessage(unusedPayloadHeader, peerNodeId, std::move(msgBuf));
 }
 
 CHIP_ERROR SecureSessionMgr::SendMessage(PayloadHeader & payloadHeader, NodeId peerNodeId, System::PacketBufferHandle msgBuf,
                                          EncryptedPacketBufferHandle * bufferRetainSlot)
 {
-    return SendMessage(payloadHeader, peerNodeId, std::move(msgBuf), bufferRetainSlot, false);
+    PacketHeader ununsedPacketHeader;
+    return SendMessage(payloadHeader, ununsedPacketHeader, peerNodeId, std::move(msgBuf), bufferRetainSlot,
+                       EncryptionState::kPayloadIsUnencrypted);
 }
 
-CHIP_ERROR SecureSessionMgr::SendMessage(EncryptedPacketBufferHandle msgBuf, EncryptedPacketBufferHandle * bufferRetainSlot)
+CHIP_ERROR SecureSessionMgr::SendEncryptedMessage(EncryptedPacketBufferHandle msgBuf,
+                                                  EncryptedPacketBufferHandle * bufferRetainSlot)
 {
+    VerifyOrReturnError(!msgBuf.IsNull(), CHIP_ERROR_INVALID_ARGUMENT);
+
+    uint16_t headerSize = 0;
+    PacketHeader packetHeader;
+    ReturnErrorOnFailure(packetHeader.Decode(msgBuf->Start(), msgBuf->DataLength(), &headerSize));
+
+    VerifyOrReturnError(packetHeader.GetDestinationNodeId().HasValue(), CHIP_ERROR_INVALID_DESTINATION_NODE_ID);
+    NodeId peerNodeId = packetHeader.GetDestinationNodeId().Value();
+
+    // Advancing the start to encrypted header, since the transport will attach the packet header on top of it
+    msgBuf->SetStart(msgBuf->Start() + headerSize);
+
     PayloadHeader payloadHeader;
+    ReturnErrorOnFailure(SendMessage(payloadHeader, packetHeader, peerNodeId, std::move(msgBuf), bufferRetainSlot,
```

Comment:
why isn't this just `return SendMessage(...);` ?

Reply:
That's a good point. I think some rework (as part of addressing comments) got rid of some intermediate code.

Fixing it here: https://github.com/project-chip/connectedhomeip/pull/4317/files#diff-95eaad1efd04391d106529cad37a61fd4a010a561587a6ad5be004d00b22bfdeR126